src/entities/swap/swapV2/index.ts

```diff
@@ -0,0 +1,339 @@
+import { TokenAmount } from '../../tokenAmount';
+import { SingleSwap, SwapKind, BatchSwapStep, Hex } from '../../../types';
+import {
+    abs,
+    BALANCER_QUERIES,
+    DEFAULT_USERDATA,
+    DEFAULT_FUND_MANAGMENT,
+    ZERO_ADDRESS,
+    NATIVE_ADDRESS,
+    VAULT,
+} from '../../../utils';
+import {
+    Address,
+    createPublicClient,
+    encodeFunctionData,
+    getContract,
+    http,
+} from 'viem';
+import { balancerQueriesAbi, vaultV2Abi } from '../../../abi';
+import cloneDeep from 'lodash.clonedeep';
+import { Path } from '../types';
+import { PathWithAmount } from '../pathWithAmount';
+import { getInputAmount, getOutputAmount } from '../pathHelpers';
+
+// A Swap can be a single or multiple paths
+export class SwapV2 {
+    public constructor({
+        chainId,
+        paths,
+        swapKind,
+    }: { chainId: number; paths: Path[]; swapKind: SwapKind }) {
+        if (paths.length === 0)
+            throw new Error('Invalid swap: must contain at least 1 path.');
+
+        this.paths = paths.map(
+            (p) =>
+                new PathWithAmount(
+                    chainId,
+                    p.tokens,
+                    p.pools,
+                    p.inputAmountRaw,
+                    p.outputAmountRaw,
+                ),
+        );
+
+        // paths with immutable pool balances
+        this.pathsImmutable = cloneDeep(this.paths);
```

Comment:
I feel like we won't need this pathsImmutable. You could also remove cloneDeep dependency as I believe this is the only place we use it.